### ISSUE TITLE
[F-B-4] Model Family - TCN & Gradient Boosting Implementations

### ISSUE THREAD

**Issue by siven76** [siven76/PA-AI#13]
## Feature Description
Implement core machine learning model architectures including Temporal Convolutional Networks (TCN) with pattern-flag channels and Gradient Boosted Trees for binary classification of trading signals.

## Acceptance Criteria
- [ ] Temporal Convolutional Network (TCN) implementation
- [ ] Pattern-flag channels integration
- [ ] Gradient Boosted Trees (XGBoost/LightGBM) implementation
- [ ] Binary classification for long/short/hold signals
- [ ] Model serialization and loading
- [ ] Feature importance analysis
- [ ] Hyperparameter optimization support
- [ ] GPU acceleration for TCN training

## Technical Requirements
- **TCN Framework**: PyTorch with custom TCN layers
- **Gradient Boosting**: XGBoost and LightGBM implementations
- **Input Features**: OHLCV + technical indicators + pattern flags
- **Output**: Binary/ternary classification (long/short/hold)
- **Performance**: Fast inference (<50ms) and efficient training

## TCN Architecture
```python
class PriceActionTCN(nn.Module):
    def __init__(self, 
                 input_size: int,
                 num_channels: List[int],
                 kernel_size: int = 3,
                 dropout: float = 0.2):
        super().__init__()
        
        # Input channels: OHLCV (5) + indicators (N) + patterns (M)
        self.tcn = TemporalConvNet(
            num_inputs=input_size,
            num_channels=num_channels,
            kernel_size=kernel_size,
            dropout=dropout
        )
        
        # Classification head
        self.classifier = nn.Sequential(
            nn.Linear(num_channels[-1], 64),
            nn.ReLU(),
            nn.Dropout(dropout),
            nn.Linear(64, 3)  # long/short/hold
        )
    
    def forward(self, x):
        # x shape: (batch, sequence, features)
        tcn_out = self.tcn(x.transpose(1, 2))  # TCN expects (batch, features, sequence)
        logits = self.classifier(tcn_out[:, :, -1])  # Use last timestep
        return logits
```

## Pattern-Flag Channels
```python
# Enhanced input with pattern flags
input_features = {
    'ohlcv': ['open', 'high', 'low', 'close', 'volume'],
    'indicators': ['rsi_14', 'sma_20', 'bb_upper', 'bb_lower', 'macd'],
    'patterns': ['hammer', 'doji', 'engulfing', 'pin_bar', 'inside_bar'],
    'pattern_confidence': ['hammer_conf', 'doji_conf', 'engulfing_conf']
}

# Total input size: 5 + 5 + 5 + 3 = 18 features per timestep
```

## Gradient Boosting Implementation
```python
class GradientBoostingSignal:
    def __init__(self, model_type='xgboost'):
        self.model_type = model_type
        if model_type == 'xgboost':
            self.model = xgb.XGBClassifier(
                objective='multi:softprob',
                num_class=3,
                max_depth=6,
                learning_rate=0.1,
                n_estimators=100,
                subsample=0.8,
                colsample_bytree=0.8
            )
        elif model_type == 'lightgbm':
            self.model = lgb.LGBMClassifier(
                objective='multiclass',
                num_class=3,
                max_depth=6,
                learning_rate=0.1,
                n_estimators=100
            )
    
    def prepare_features(self, df):
        """Convert time series to feature matrix"""
        features = []
        lookback = 20
        
        for i in range(lookback, len(df)):
            window = df.iloc[i-lookback:i]
            
            # Price features
            price_features = [
                window['close'].pct_change().mean(),
                window['high'].max() / window['low'].min() - 1,
                (window['close'].iloc[-1] - window['open'].iloc[0]) / window['open'].iloc[0]
            ]
            
            # Pattern features (aggregated over window)
            pattern_features = [
                window['hammer'].sum(),
                window['doji'].sum(),
                window['engulfing'].sum()
            ]
            
            # Technical indicators (latest values)
            indicator_features = [
                window['rsi_14'].iloc[-1],
                window['sma_20'].iloc[-1],
                window['bb_upper'].iloc[-1] - window['bb_lower'].iloc[-1]
            ]
            
            features.append(price_features + pattern_features + indicator_features)
        
        return np.array(features)
```

## Model Comparison Framework
```python
class ModelComparison:
    def __init__(self):
        self.models = {
            'tcn': PriceActionTCN(input_size=18, num_channels=[32, 64, 32]),
            'xgboost': GradientBoostingSignal('xgboost'),
            'lightgbm': GradientBoostingSignal('lightgbm')
        }
    
    def evaluate_models(self, X_train, y_train, X_test, y_test):
        results = {}
        
        for name, model in self.models.items():
            # Train model
            if name == 'tcn':
                results[name] = self.train_tcn(model, X_train, y_train, X_test, y_test)
            else:
                results[name] = self.train_gb(model, X_train, y_train, X_test, y_test)
        
        return results
```

## Performance Requirements
- **Training Time**: TCN <2 hours, GB <30 minutes on single GPU/CPU
- **Inference Speed**: <50ms per prediction on CPU
- **Memory Usage**: <4GB RAM during training
- **Accuracy**: >60% on out-of-sample test data
- **Sharpe Ratio**: >1.0 on backtested signals

## Definition of Done
- [ ] TCN implementation with PyTorch
- [ ] Pattern-flag channel integration
- [ ] XGBoost and LightGBM implementations
- [ ] Feature engineering pipeline
- [ ] Model comparison framework
- [ ] Hyperparameter optimization
- [ ] Model serialization/deserialization
- [ ] Performance benchmarking
- [ ] Unit tests for all model components
- [ ] Integration with F-B-3 training pipeline
